### PR TITLE
Stream Deployment: UX improvements

### LINKED ISSUE #633: Improve stream-deployment UX
Continuing from spring-cloud/spring-cloud-dataflow-ui#515, as a developer, I'd like to implement the new stream-deployment UX.

**Acceptance:**
- [Original requirements](https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/515#issuecomment-364801670) are validated
- Implement mutual exclusivity between global vs. app-specific property overrides
- Document new UX in SCDF-core docs (replace screenshots)
- Assumptions captured
- Improve test coverage

### LINKED ISSUE #647: Implement modal-window for properties selection from stream deployment
This is the same properties modal window used in the Flo canvas; we would like to reuse that.

### PR DESCRIPTION
**Pull request by oodamien**
Work In Progress
Resolves #633
Resolves #647

### CODE REVIEW

**Comment by oodamien** [comment]
TODO:
- UI
- title app properties modal

**Comment by sabbyanandan** [comment]
@ilayaperumalg: While @ghillert is reviewing the implementation, it'd be great if you can review it from the functional perspective. Apps, deployer, and global properties - all of that wrapped into the builder-form. Let us know what you think.

**Comment by oodamien** [comment]
DO NOT MERGE

File: ui/src/app/streams/stream-definitions/deployment-properties/deployment-properties.component.ts [hunk 1]
```diff
@@ -66,17 +66,17 @@ export class DeploymentPropertiesComponent implements OnInit, OnDestroy {
...
-        .pipe(takeUntil(this.ngUnsubscribe$))
-        .subscribe((platforms: Platform[]) => {
-          this.platforms = platforms;
-        });
-      }
-    });
+      .pipe(takeUntil(this.ngUnsubscribe$))
+      .subscribe(featureInfo => {
+        this.skipperEnabled = featureInfo.skipperEnabled;
+        if (this.skipperEnabled) {
```
Comment (ilayaperumalg) [inline ui/src/app/streams/stream-definitions/deployment-properties/deployment-properties.component.ts]: In terms of Skipper related properties, It looks like we only allow platform name to be specified via deployment properties. Are the properties such as package version, repo name are planned to be available as next set of features? [This Shell Command](https://github.com/spring-cloud/spring-cloud-dataflow/blob/15095222f7693b04573ae5aae678e0e991367758/spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/command/skipper/SkipperStreamCommands.java#L91) contains the available options for Skipper deployment properties.

File: ui/src/app/streams/stream-deploy/stream-deploy.service.ts [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+          versions: app.versions,
+          options: null,
+          optionsState: {
+            isLoading: false,
+            isOnError: false
+          }
+        }));
+
+        // Deployers
+        config.deployers = [
```
Comment (ilayaperumalg) [inline ui/src/app/streams/stream-deploy/stream-deploy.service.ts]: Do we want to expose these deployer properties which are specific to the target platform (Local, CF and Kubernetes)? Though only the applicable properties would work on the target platform, not sure if we want to have a common list of these properties here.

File: ui/src/app/streams/stream-deploy/stream-deploy.component.ts [hunk 3]
```diff
@@ -21,66 +27,352 @@ import { takeUntil } from 'rxjs/operators';
...
+    if (this.state.view === 'builder') {
+      const deployers: FormArray = value.builder.get('deployers') as FormArray;
+      const appsVersion: FormGroup = value.builder.get('appsVersion') as FormGroup;
+      const global: FormArray = value.builder.get('global') as FormArray;
+      const specificPlatform: FormArray = value.builder.get('specificPlatform') as FormArray;
+
+      // Platform
+      if (!isEmpty(value.builder.get('platform'))) {
+        result.push({
+          key: 'deployer.platform',
```
Comment (ilayaperumalg) [inline ui/src/app/streams/stream-deploy/stream-deploy.component.ts]: The key `deployer.platform` looks incorrect as it needs to be `spring.cloud.dataflow.skipper.platformName`

**Comment by ghillert** [comment]
@oodamien Seeing merge conflict. Please rebase.

**Comment by ghillert** [comment]
Running into UI issue when running in `classic-mode` (non-Skipper)

image

@sabbyanandan @markpollack To simplify things, should we just use the old page for `classic-mode` and use the the new page only for `skipper-mode`?

**Comment by sabbyanandan** [comment]
As discussed in chat, we can simply just do the new rendering _only_ for Skipper. The Classic-mode can continue to use the old page. We can decide the redirect based on the Skipper feature-flag available in the system.

**Comment by ghillert** [comment]
@oodamien Could we maybe make the first column a bit narrower? We could probably make the column 50% smaller. 

Also, the *static* non-changing properties should probably not be text-input field (I think that may confuse users)

Another one - I know this is probably a bit tricky, could you improve the `tabbing` behavior (using the tab-key to go from field to field)? 

image

E.g. when entering a new property, pressing the `tab` should jump to the next `value` text input field.

**Comment by ghillert** [comment]
We should probably relax the property name restrictions - I think `my-value` and `my_value` are legal.

image

**Comment by ghillert** [comment]
Maybe add a download icon to the button to highlight that the file will be downloaded: 

image

**Comment by ghillert** [comment]
Low-hanging fruit and I think will be quite useful: Can you add a button similar to what we have on the About Details page to copy the properties into the clipboard?

image

**Comment by ghillert** [comment]
image

It would be nice and convenient if we had also the option to populate a text-area via the clipboard. Currently, this line makes no sense: `This will be used to populate the text area above.`

File: ui/src/app/streams/stream-deploy/properties-debug/properties-debug.component.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  <div class="property" *ngFor="let val of raw" [class.invalid]="val.status == 'invalid'">
+    <span *ngIf="val.status == 'invalid'" class="fa fa-warning">
+    </span>
+    <span class="key">{{ val.key }}</span>
+    <span class="equal"> = </span>
+    <span class="value">{{ val.value }}</span>
+  </div>
+</div>
+<div class="alert alert-info" style="display: inline-block" *ngIf="raw && raw.length == 0">
+  No property define
```
Comment (ghillert) [inline ui/src/app/streams/stream-deploy/properties-debug/properties-debug.component.html]: Change to `No properties defined`

File: ui/src/app/streams/stream-deploy/stream-deploy.component.ts [hunk 5]
```diff
@@ -93,64 +385,76 @@ export class StreamDeployComponent implements OnInit, OnDestroy {
...
-    const file: File = event.target.files[0];
-    const reader: FileReader = new FileReader();
-    const _form = this.form;
-    reader.onloadend = () => {
-      _form.patchValue({deploymentProperties: reader.result});
-    };
-    reader.readAsText(file);
+  exportBuilderToFile(value: { builder: FormGroup, builderAppsProperties: any, file: FormGroup, config: StreamDeployConfig }) {
+    const propertiesString = this.toRaw(value).map((a) => `${a.key}=${a.value}`).join('\n');
+    const filename = `${value.config.id}.${new Date().getTime()}.txt`;
```
Comment (ghillert) [inline ui/src/app/streams/stream-deploy/stream-deploy.component.ts]: Could we use `momentjs` to generate a more user-friendly time (ISO 8601 format)? A unix timestamp is a bit hard on the eye ;-)

E.g.:   `test_2018-03-14T162355.txt`

**Comment by ghillert** [comment]
image

If nothing is filled in, I would use the plural `Cores`. If we are fancy, we can make it singular if a `1` is used as a value :-)

Also, change `Mb` to `MB` (Mb would theoretically mean Megabit)

**Comment by sabbyanandan** [comment]
I moved copy-to-clipboard comment to a different story (spring-cloud/spring-cloud-dataflow-ui#663).

@oodamien: Let's address the width and the other cosmetic issues in this PR.

**Comment by ghillert** [comment]
Merge with commit 5697ceb72218dd2313497a428b0a67867a33a10d

**Comment by ghillert** [comment]
Created follow-up issues:

https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/664
https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/665
https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/666
https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/667

**Pull request closed without merging**